Fix Url Rewrite not generated for stores that have different url key and adds the deprecated message to CategoryBasedProductRewriteGenerator class

TuVanDev opened the pull request:
### Description (*)
This pull request includes fixes for #36295 and adds the deprecated message to `Magento\CatalogUrlRewrite\Model\CategoryBasedProductRewriteGenerator` class due to it is no longer used anywhere after the change in commit https://github.com/magento/magento2/commit/c7d6324fe84766ddd9ab9710f19b5923fcbac09a (Magento 2.2.0) which `\Magento\CatalogUrlRewrite\Model\CategoryProductUrlPathGenerator` was introduced to replace `Magento\CatalogUrlRewrite\Model\CategoryBasedProductRewriteGenerator`.
1. Fixes #36295
2. Fix another related issue to #36295: URL Rewrite not generated for stores that have different URL key when updating category products/product in the `store` scope (changing product's categories under another store)

3. Add the @deprecated tag to the `Magento\CatalogUrlRewrite\Model\CategoryBasedProductRewriteGenerator` class and indicate that the `Magento\CatalogUrlRewrite\Model\CategoryProductUrlPathGenerator` class should be used.

**Assumption**:
- Magento 2 (affected to all versions) with 2 store views
- Configurations -> Catalog -> Catalog tab -> Search Engine Optimization -> Generate "category/product" URL Rewrites = **Yes**
- Product A used the default value of url key in the store view 1
- Product A used a different url key (do not use default value) in the store view 2

**Behavior**:

Updating (add/remove) product categories for product A in one of the below scopes:
  - Global scope
  - **Store view 1**

**Expected result**: Url rewrite regenerate for both store view 1 and store view 2.

**Actually result**: Url rewrite only generates for store view 1 but does not regenerate for store view 2.

<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)

1. Fixes magento/magento2#36295

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ...
2. ...

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

TuVanDev committed:
Remove CategoryBasedProductRewriteGenerator class as it is redundant since commit #c7d6324

TuVanDev committed:
#36295: Fix Url rewrites not regenrated for stores that have different url-key

Viper9x commented:
@magento give me test instance

Viper9x commented:
@magento give me 2.4-develop instance

Den4ik reviewed (changes requested):
Hi @Viper9x. Thanks for you work.
Removing of the class is backward incomparable change and should be approved by architects. I believe that use of `@deprecate` message more useful for fast delivery.
You mentioned https://github.com/magento/magento2/commit/c7d6324fe84766ddd9ab9710f19b5923fcbac09a that replace object manager usage with injection to constructor of `\Magento\CatalogUrlRewrite\Model\CategoryProductUrlPathGenerator`. It's not `\Magento\CatalogUrlRewrite\Model\CategoryBasedProductRewriteGenerator`

TuVanDev committed:
Revert "Remove CategoryBasedProductRewriteGenerator class as it is redundant since commit #c7d6324"

This reverts commit d84fded5b0b9dffabb01468e9805d8f4ac0b63a1.

Viper9x commented:
Hi @Den4ik.
I only say that `Magento\CatalogUrlRewrite\Model\CategoryBasedProductRewriteGenerator` is **no longer used** after the change in commit https://github.com/magento/magento2/commit/c7d6324fe84766ddd9ab9710f19b5923fcbac09a (Magento 2.2.0).
More explanation about that: commit https://github.com/magento/magento2/commit/c7d6324fe84766ddd9ab9710f19b5923fcbac09a changed from using `Magento\CatalogUrlRewrite\Model\CategoryBasedProductRewriteGenerator` to `\Magento\CatalogUrlRewrite\Model\CategoryProductUrlPathGenerator`, also replace object manager usage with injection to constructor. In Magento_CatalogUrlRewrite module, only `Magento\CatalogUrlRewrite\Observer\UrlRewriteHandler` class
uses `Magento\CatalogUrlRewrite\Model\CategoryBasedProductRewriteGenerator`, and after that commit, `Magento\CatalogUrlRewrite\Model\CategoryBasedProductRewriteGenerator` is not used anymore.

Do you think adding ` * @deprecated 101.0.0 use \Magento\CatalogUrlRewrite\Observer\UrlRewriteHandler instead.` to `Magento\CatalogUrlRewrite\Model\CategoryBasedProductRewriteGenerator` class's docBlock is fine?

Thanks for your support.

TuVanDev committed:
Merge branch '2.4-develop' into fix-url-rewrite-not-generated-for-stores-that-have-different-url-key

TuVanDev committed:
Add deprecated message for CategoryBasedProductRewriteGenerator class

Magento\CatalogUrlRewrite\Model\CategoryBasedProductRewriteGenerator is no longer used after the change in commit c7d6324

Viper9x commented:
Hi @Den4ik I've added the deprecated message to `Magento\CatalogUrlRewrite\Model\CategoryBasedProductRewriteGenerator` class due to it is no longer used after the change in commit c7d6324
Please check, thanks.

hostep commented:
@Viper9x: it looks like a commit made it in the 2.4-develop branch by the core team: https://github.com/magento/magento2/commit/e46f719559924e703ef3c7fdc728ec1b232a7509

I have the feeling it might solve the same issue you are trying to solve, could you investigate that?
If it does solve the same issue, this PR can be closed.
If it doesn't solve the same issue, you'll have to update your PR so that no merge conflicts exist anymore.

Thanks! 🙂

TuVanDev committed:
Merge branch origin/2.4-develop

TuVanDev committed:
#36925: Correct deprecated message for CategoryBasedProductRewriteGenerator class

Viper9x commented:
@hostep: Thanks for let me know about the update related to my PR. I've investigated and tested for sure.

The commit https://github.com/magento/magento2/commit/e46f719559924e703ef3c7fdc728ec1b232a7509 resolve related issue but not the issue I'm trying to solve.

Detail:
- The commit https://github.com/magento/magento2/commit/e46f719559924e703ef3c7fdc728ec1b232a7509: fix URL Rewrite not generated for stores that have different url key when updating category products/ product in `global` scope, it does not fix when updating category products/ product in `store` scope
- My commits in this PR: fix URL Rewrite not generated for stores that have different url key when updating category products/ product in both `global` and `store` scope.

I have been updated my PR with the lastest changes from `2.4-develop branch`, so my PR won't be conflict with other.

Thanks Pieter Hoste

hostep commented:
@magento run all tests

Den4ik commented:
@Viper9x Thanks for update. Could you please check failed Unit and Static tests

TuVanDev committed:
Merge branch 'magento:2.4-develop' into fix-url-rewrite-not-generated-for-stores-that-have-different-url-key

TuVanDev committed:
magento#36295: Update Unit Test

Update Unit Test to be compatible with the new changes in ProductUrlRewriteGenerator class

Viper9x commented:
@magento run all tests

TuVanDev committed:
magento#36295: Update Unit Test - remove unused variable

Update Unit Test to be compatible with the new changes in ProductUrlRewriteGenerator class

Viper9x commented:
@magento run all tests

Viper9x commented:
@Den4ik I've updated Unit Test to be compatible with the new changes in `Magento\CatalogUrlRewrite\Model\ProductUrlRewriteGenerator` class.

For Static Tests, there are warnings does not relate to my changes and it requires much time to resolve many warnings caused by "@deprecated tag MUST be explained". So, can we ignore these warnings?

engcom-Delta commented:
Hi @Viper9x,
We have tried to reproduce the issue, according the steps which are mentioned in the [36295](https://github.com/magento/magento2/issues/36295) this issue is not reproducible

Preconditions:
1. Magento 2.4-develop
2. PHP 8.1

To reproduce, we followed the steps below
1. Configurations -> Catalog [Catalog tab] -> Search Engine Optimization -> Generate "category/product" URL Rewrites = Yes
<img width="1401" alt="Screenshot 2023-02-17 at 9 06 05 PM" src="https://user-images.githubusercontent.com/51681379/219704319-18df3ba2-6b30-41fc-a161-d26572a78a39.png">

2. Create a new store view and assign it to a current Store (There should be 2 store views)
<img width="1402" alt="Screenshot 2023-02-17 at 9 07 49 PM" src="https://user-images.githubusercontent.com/51681379/219704482-cd2de281-1d7f-4b3e-9ee8-805bc80f3848.png">

3. Create 2 new categories under root category.
<img width="413" alt="Screenshot 2023-02-17 at 9 09 18 PM" src="https://user-images.githubusercontent.com/51681379/219704591-e4e83107-ae26-4d78-b2d9-86bfe34870e3.png">

4. Create a new product and assign a single category to a product.
<img width="1142" alt="Screenshot 2023-02-17 at 9 11 12 PM" src="https://user-images.githubusercontent.com/51681379/219704727-7d4e803f-7723-4acf-bf96-43f75bd680a3.png">

5. Edit a product under a first store scope and unselect "Use Default Value" for a "Url key" field in "Search Engine 
Optimization" tab.
<img width="723" alt="Screenshot 2023-02-17 at 9 11 45 PM" src="https://user-images.githubusercontent.com/51681379/219704870-0a3a36c8-4215-4586-96d9-366f3c9fe19a.png">
<img width="1083" alt="Screenshot 2023-02-17 at 9 12 51 PM" src="https://user-images.githubusercontent.com/51681379/219705089-f27166de-a906-4e89-929a-48c806291129.png">

6. Edit a product under "All store views" scope and assign the 2nd category.
<img width="749" alt="Screenshot 2023-02-17 at 9 14 48 PM" src="https://user-images.githubusercontent.com/51681379/219705277-14795f50-4e9d-40b1-b083-5e69b3943e2f.png">
<img width="1036" alt="Screenshot 2023-02-17 at 9 15 18 PM" src="https://user-images.githubusercontent.com/51681379/219705392-8068ee8f-90f6-443a-891d-972fe390ba22.png">
7. Marketing ->url rewrites
<img width="1397" alt="Screenshot 2023-02-17 at 9 18 48 PM" src="https://user-images.githubusercontent.com/51681379/219705676-4f70c3f2-0478-4e0e-b302-7f1b8f2656c5.png">

In the  above screenshot the url rewrite is generated for a 2nd assigned category in the first store view. hence currently this issue is not exist.

If we are missed anything, please let us know.

Thanks.

Den4ik commented:
Hi @engcom-Delta 
Thanks for good steps description. Let me check it locally too.

Viper9x commented:
Hey @engcom-Delta , thanks for joining this pull request.
Issue #36295 was resolved since a commit was made in the 2.4-develop branch recently by the core team: https://github.com/magento/magento2/commit/e46f719559924e703ef3c7fdc728ec1b232a7509

My PR provide the fix for both issue #36295 and another `related` issue which has never been reported yet (will describe below). I actually created this pull request a week ahead of the [fix](https://github.com/magento/magento2/commit/e46f719559924e703ef3c7fdc728ec1b232a7509) from the core team for ACP2E-1304 which is also resolved #36295. I've replied @hostep for that above.
Even though I'm already mentioned it in the **behavior** section but it looks not clear since issue #36295 was resolved. I'll update the fixes list now.

Another related issue I mentioned is: "URL Rewrite not generated for stores that have different URL key when updating category products/product in the `store` scope: https://github.com/magento/magento2/pull/36360#issuecomment-1312405959

So, we need one more testing scenario to reproduce the issue:
1. Install a new Magento instance.
2. Create a new store view and assign it to a current store (there should be 2 store views)
3. Create 2 new categories under the root category.
4. Create a new product and assign a single category to a product.
5. Edit a product under a first store scope and unselect "Use Default Value" for a "Url key" field in "Search Engine Optimization" tab.
6. Edit a product under the second `store scope` and assign the 2nd category.

Note: steps 1 through 5 are identical to #36295 but step 6 is different: edit a product under another `store view scope` instead of All store views scope. P/S: I can share another testing scenario that has fewer steps but this one is almost the same as  #36295, so we can utilize it.
cc @Den4ik

Viper9x commented:
Oh, just realized that I've provided enough all information in **Assumption**, **Behavior**, **Expected result**, **Actually result** before.

engcom-Delta commented:
Hi @Viper9x 
Thank you for sharing [this](https://github.com/magento/magento2/pull/36360#issuecomment-1435480876) information.

engcom-Delta committed:
Merge remote-tracking branch 'magento/2.4-develop' into fix-url-rewrite-not-generated-for-stores-that-have-different-url-key

engcom-Delta commented:
@magento run all tests

engcom-Delta commented:
I am currently working on the Priority task. Once it is completed, I will resume working on this PR.

Thanks

engcom-Delta committed:
PR:36360 fix for Static Tests failures

engcom-Delta commented:
@magento run all tests

engcom-Delta commented:
@magento run Functional Tests CE

engcom-Delta commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

engcom-Delta committed:
Merge remote-tracking branch 'magento/2.4-develop' into fix-url-rewrite-not-generated-for-stores-that-have-different-url-key

engcom-Delta commented:
@magento run all tests

engcom-Delta committed:
Merge remote-tracking branch 'magento/2.4-develop' into fix-url-rewrite-not-generated-for-stores-that-have-different-url-key

engcom-Delta committed:
PR-36360: fix for functional test failure

engcom-Delta commented:
@magento run all tests

engcom-Delta committed:
Removing .credential file

engcom-Delta commented:
@magento run all tests

engcom-Delta committed:
fix for functional test failures

engcom-Delta commented:
@magento run all tests

engcom-Delta commented:
@magento run Integration Tests

engcom-Lima committed:
Merge branch '2.4-develop' into fix-url-rewrite-not-generated-for-stores-that-have-different-url-key

engcom-Delta commented:
@magento run all tests

engcom-Delta committed:
Merge remote-tracking branch 'magento/2.4-develop' into fix-url-rewrite-not-generated-for-stores-that-have-different-url-key

engcom-Delta committed:
fix for integration test failure

engcom-Delta commented:
@magento run all tests

engcom-Delta commented:
@magento run WebAPI Tests

cptX commented:
I have magento 2.4.5 and magento 2.4.6. When I update the url key of a category that has subcategories the url key is not updated in the frontend, although it is saved in the category settings. Also in url rewrites the new url is not visible at all. Could this be related to the issue you try to solve here?

Viper9x commented:
@cptX I don't have time to check that issue, could you apply the patch for this pull request to see if it resolves your issue?
Here is the patch for this pull request:
```
diff --git a/vendor/magento/module-catalog-url-rewrite/Model/CategoryBasedProductRewriteGenerator.php b/vendor/magento/module-catalog-url-rewrite/Model/CategoryBasedProductRewriteGenerator.php
index 105ca290..8e599f91 100644
--- a/vendor/magento/module-catalog-url-rewrite/Model/CategoryBasedProductRewriteGenerator.php
+++ b/vendor/magento/module-catalog-url-rewrite/Model/CategoryBasedProductRewriteGenerator.php
@@ -11,8 +11,9 @@ use Magento\Catalog\Model\Product\Visibility;

 /**
  * Class ProductUrlRewriteGenerator
- * @package Magento\CatalogUrlRewrite\Model
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ * @deprecated 101.0.0 please use \Magento\CatalogUrlRewrite\Model\CategoryProductUrlPathGenerator instead.
+ * @see https://github.com/magento/magento2/commit/c7d6324fe84766ddd9ab9710f19b5923fcbac09a
  */
 class CategoryBasedProductRewriteGenerator
 {
diff --git a/vendor/magento/module-catalog-url-rewrite/Model/ProductUrlRewriteGenerator.php b/vendor/magento/module-catalog-url-rewrite/Model/ProductUrlRewriteGenerator.php
index f5e6ae9a..d1785773 100644
--- a/vendor/magento/module-catalog-url-rewrite/Model/ProductUrlRewriteGenerator.php
+++ b/vendor/magento/module-catalog-url-rewrite/Model/ProductUrlRewriteGenerator.php
@@ -6,16 +6,14 @@
 namespace Magento\CatalogUrlRewrite\Model;

 use Magento\Catalog\Model\Product;
+use Magento\Catalog\Model\Product\Visibility;
 use Magento\CatalogUrlRewrite\Model\Product\CanonicalUrlRewriteGenerator;
 use Magento\CatalogUrlRewrite\Model\Product\CategoriesUrlRewriteGenerator;
 use Magento\CatalogUrlRewrite\Model\Product\CurrentUrlRewritesRegenerator;
 use Magento\CatalogUrlRewrite\Service\V1\StoreViewService;
 use Magento\Framework\App\ObjectManager;
-use Magento\Catalog\Model\Product\Visibility;

 /**
- * Class ProductUrlRewriteGenerator
- * @package Magento\CatalogUrlRewrite\Model
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
  */
 class ProductUrlRewriteGenerator
@@ -23,10 +21,11 @@ class ProductUrlRewriteGenerator
     /**
      * Entity type code
      */
-    const ENTITY_TYPE = 'product';
+    public const ENTITY_TYPE = 'product';

     /**
      * @deprecated 100.1.0
+     * @see https://github.com/magento/magento2/commit/6729b6e01368248abc33300208eb292c95050203
      * @var \Magento\CatalogUrlRewrite\Service\V1\StoreViewService
      */
     protected $storeViewService;
@@ -34,41 +33,48 @@ class ProductUrlRewriteGenerator
     /**
      * @var \Magento\Catalog\Model\Product
      * @deprecated 100.1.0
+     * @see https://github.com/magento/magento2/commit/6729b6e01368248abc33300208eb292c95050203
      */
     protected $product;

     /**
      * @deprecated 100.1.0
+     * @see https://github.com/magento/magento2/commit/6729b6e01368248abc33300208eb292c95050203
      * @var \Magento\CatalogUrlRewrite\Model\Product\CurrentUrlRewritesRegenerator
      */
     protected $currentUrlRewritesRegenerator;

     /**
      * @deprecated 100.1.0
+     * @see https://github.com/magento/magento2/commit/6729b6e01368248abc33300208eb292c95050203
      * @var \Magento\CatalogUrlRewrite\Model\Product\CategoriesUrlRewriteGenerator
      */
     protected $categoriesUrlRewriteGenerator;

     /**
      * @deprecated 100.1.0
+     * @see https://github.com/magento/magento2/commit/6729b6e01368248abc33300208eb292c95050203
      * @var \Magento\CatalogUrlRewrite\Model\Product\CanonicalUrlRewriteGenerator
      */
     protected $canonicalUrlRewriteGenerator;

     /**
      * @deprecated 100.1.0
+     * @see https://github.com/magento/magento2/commit/6729b6e01368248abc33300208eb292c95050203
      * @var \Magento\CatalogUrlRewrite\Model\ObjectRegistryFactory
      */
     protected $objectRegistryFactory;

     /**
      * @deprecated 100.1.0
+     * @see https://github.com/magento/magento2/commit/6729b6e01368248abc33300208eb292c95050203
      * @var \Magento\CatalogUrlRewrite\Model\ObjectRegistry
      */
     protected $productCategories;

     /**
      * @deprecated 100.1.0
+     * @see https://github.com/magento/magento2/commit/6729b6e01368248abc33300208eb292c95050203
      * @var \Magento\Store\Model\StoreManagerInterface
      */
     protected $storeManager;
@@ -106,13 +112,14 @@ class ProductUrlRewriteGenerator
      * Retrieve Delegator for generation rewrites in different scopes
      *
      * @deprecated 100.1.4
+     * @see https://github.com/magento/magento2/commit/b2ce2a37d921b5ad88fc38663fc0ff3dd6c582d1
      * @return ProductScopeRewriteGenerator|mixed
      */
     private function getProductScopeRewriteGenerator()
     {
         if (!$this->productScopeRewriteGenerator) {
             $this->productScopeRewriteGenerator = ObjectManager::getInstance()
-            ->get(ProductScopeRewriteGenerator::class);
+                ->get(ProductScopeRewriteGenerator::class);
         }

         return $this->productScopeRewriteGenerator;
@@ -131,23 +138,19 @@ class ProductUrlRewriteGenerator
             return [];
         }

-        $storeId = $product->getStoreId();
-
         $productCategories = $product->getCategoryCollection()
             ->addAttributeToSelect('url_key')
             ->addAttributeToSelect('url_path');

-        $urls = $this->isGlobalScope($storeId)
-            ? $this->generateForGlobalScope($productCategories, $product, $rootCategoryId)
-            : $this->generateForSpecificStoreView($storeId, $productCategories, $product, $rootCategoryId);
-
-        return $urls;
+        // Generate url rewrites for all store views to ensure store views with different url-key are generated as well.
+        return $this->generateForGlobalScope($productCategories, $product, $rootCategoryId);
     }

     /**
      * Check is global scope
      *
      * @deprecated 100.1.4
+     * @see https://github.com/magento/magento2/commit/b2ce2a37d921b5ad88fc38663fc0ff3dd6c582d1
      * @param int|null $storeId
      * @return bool
      */
@@ -160,6 +163,7 @@ class ProductUrlRewriteGenerator
      * Generate list of urls for global scope
      *
      * @deprecated 100.1.4
+     * @see https://github.com/magento/magento2/commit/b2ce2a37d921b5ad88fc38663fc0ff3dd6c582d1
      * @param \Magento\Framework\Data\Collection $productCategories
      * @param \Magento\Catalog\Model\Product|null $product
      * @param int|null $rootCategoryId
@@ -178,6 +182,7 @@ class ProductUrlRewriteGenerator
      * Generate list of urls for specific store view
      *
      * @deprecated 100.1.4
+     * @see https://github.com/magento/magento2/commit/b2ce2a37d921b5ad88fc38663fc0ff3dd6c582d1
      * @param int $storeId
      * @param \Magento\Framework\Data\Collection $productCategories
      * @param Product|null $product
@@ -195,7 +200,10 @@ class ProductUrlRewriteGenerator
     }

     /**
+     * Checking the category is generating correctly
+     *
      * @deprecated 100.1.4
+     * @see https://github.com/magento/magento2/commit/b2ce2a37d921b5ad88fc38663fc0ff3dd6c582d1
      * @param \Magento\Catalog\Model\Category $category
      * @param int $storeId
      * @return bool

```

cptX commented:
@Viper9x I have the setting "Use Categories Path for Product URLs" to "No". My problem is only with the category url rewrites. Will this patch help this? I see that the above modifications affect product url rewrites...

cptX commented:
@Viper9x Unfortunately this patch didn't solve my problem. I'm still unable to create a different URL rewrite for my second store view. Rewrite URL is visible in the configuration but not working. Browser says "The page isn’t redirecting properly". As I said I have the setting "Use Categories Path for Product URLs" to "No".
I'm feeling that the issue is much deeper than what this patch is doing...

engcom-Dash commented:
Hello @TuVanDev ,

Thanks for the contributions!

Please resolve the conflict so we can proceed with this PR.

Thanks.

engcom-Dash commented:
Closing this PR since it has not been updated in a while. Please feel free to reopen if needed.

engcom-Dash commented:
Closing this PR since it has not been updated in a while. Please feel free to reopen if needed.

TuVanDev commented:
Hey @engcom-Dash,
Could you please reopen this ticket?

Den4ik commented:
@TuVanDev PR reopened. Please resolve conflict.

TuVanDev committed:
Merge branch '2.4-develop' into fix-url-rewrite-not-generated-for-stores-that-have-different-url-key

TuVanDev commented:
@Den4ik @engcom-Dash I have resolved the merge conflict. Please review the changes.
